File: server/src/routes/geolocate.ts
import { Router, Request, Response } from 'express';
import dns from 'node:dns';
import ErrorResponse from '../utils/errorResponse.js';
import '../lib/net.js'; // force ipv4 for fetch

const router = Router();

router.post('/', async (req: Request, res: Response) => {
    const { lat, lng } = req.body;

    if (!lat || !lng) {
        return res.status(400).json({ error: "Missing lat/lng" });
    }

    try {
        const url = `${process.env.NOMINATION_API_HOST}?format=jsonv2&lat=${lat}&lon=${lng}`;
        const response = await fetch(url, {
            headers: {
                "User-Agent": `slo-id-client/1.0 (${process.env.EMAIL_SIGNER})`
            }
        });
        const data = await response.json();
        const { boundingbox } = data;
        const { road, town, city, municipality, state, postcode, country_code } = data.address;
        
        if (!city && !municipality) {
            return res.status(404).json({ error: "No components found" });
        }

        return res.status(200).json({ 
            road, town, city, municipality, state, postcode, country_code, boundingbox 
        });
    } catch (err: unknown) {
        console.log("DNS default order:", dns.getDefaultResultOrder());
        return ErrorResponse("Failed to fetch geolocation", err, 500, res);
    }
});

export default router;

File: server/routers/geolocate.py
import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from config import settings

router = APIRouter(prefix="/api/geolocate", tags=["geolocate"])


class GeolocateRequest(BaseModel):
    lat: float
    lng: float


@router.post("/")
async def geolocate(body: GeolocateRequest):
    url = f"{settings.nominatim_api_host}?format=jsonv2&lat={body.lat}&lon={body.lng}"
    headers = {"User-Agent": f"slo-id/1.0 ({settings.email_signer})"}

    async with httpx.AsyncClient() as client:
        try:
            resp = await client.get(url, headers=headers, timeout=10)
            resp.raise_for_status()
        except httpx.HTTPError as e:
            raise HTTPException(status_code=502, detail=f"Nominatim error: {e}")

    data = resp.json()
    address = data.get("address", {})

    city = address.get("city") or address.get("municipality")
    if not city:
        raise HTTPException(status_code=404, detail="No location components found")

    return {
        "road":         address.get("road"),
        "town":         address.get("town"),
        "city":         city,
        "municipality": address.get("municipality"),
        "state":        address.get("state"),
        "postcode":     address.get("postcode"),
        "country_code": address.get("country_code"),
        "boundingbox":  data.get("boundingbox"),
    }
